### ISSUE TITLE
Parameters pallet

### ISSUE THREAD

**Issue by ggwpez** [paritytech/substrate#14293]
This probably should be alterable via governance. Maybe it's time we introduced a generic Parameters pallet to give a place for all these parameters to sit?

_Originally posted by @gavofyork in https://github.com/paritytech/cumulus/pull/2607#discussion_r1203905861_

**Comment by xlc** [comment]
I am going to implement this in ORML

**Comment by xlc** [comment]
https://github.com/open-web3-stack/open-runtime-module-library/pull/927

**Comment by kianenigma** [comment]
don't have anything against a central pallet per se, but this could have also been something like 

`#[pallet::storage(configurable)]` or `[pallet::storage(configurable(T::CustomOrigin))]` that would auto generate setters with the given `EnsureOrigin` impl, and `EnsureRoot` if left blank. 

If named `#[pallet::setter]`, could have made the story around `getters` and `setters` also more consistent, although I am personally still not a fan of getters to begin with (see #13365). 

So all in all, something like:

```rust
#[pallet::storage]
#[pallet::getter(fn foo)]
#[pallet::setter(T::FirstOrigin)] // `T::FirstOrigin::ensure_origin(o)?` is enough to set `Foo` via `set_foo`. 
type Foo<T: Config> = StorageValue<Value = u32>
```

**Comment by ggwpez** [comment]
>UPDATE: call index of something like above would also be hard to manage -1

Yes, unless we only have one `set_parameter` function. It could be made type-safe by using an Enum over all parameters.  
But i agree that we can just start with a pallet as well.

**Comment by xlc** [comment]
A central pallet allows a single place for all the parameters. This is super useful for governance / audit / review. No need to check every single pallet for governance calls, which each of them may have different naming convention etc.

**Comment by xlc** [comment]
One thing about variable `Get` is that it could make metadata variable, which is unexpected https://github.com/paritytech/substrate/issues/11619

**Comment by muharem** [comment]
how about having these features:
- custom origin pre key (we could use a different tracks to set different values)
- default value for a key (might be useful and existing pallets do not have to be adjusted to expect the None)
- a new attribute and trait to replace `#[pallet::constant]`, where the trait impl can return the full key for the value in the storage. The full key can be included to the metadata, and clients can fetch the value from storage or use default if the storage value is None.

what you think?

**Comment by xlc** [comment]
Every feature request must come with some use cases otherwise we will be wasting time implementing things no one is using, which we already have a few instances in Substrate.

**Comment by muharem** [comment]
I believe we already have the use cases for those features,
- for example, fellowship salary budget will be controlled by `Treasurer` origin/track, staking related configs by `StakingAdmin`;
- if we can not set a default value, the pallets has to be adjusted to handle None variant;
- if no new attribute with the full key to the storage value, there is no way for clients to know the value behind the key (now they get it from metadata as constant).

**Comment by xlc** [comment]
https://github.com/open-web3-stack/open-runtime-module-library/blob/7ac9410a8e61fc74b025354d88a0beff61152dca/parameters/src/mock.rs#L75

You have access to the key and can do whatever logics you want to control the permission.

https://github.com/open-web3-stack/open-runtime-module-library/blob/7ac9410a8e61fc74b025354d88a0beff61152dca/traits/src/parameters.rs#L14

this returns an option so user will just add unwrap_or_else to support default value


> if no new attribute with the full key to the storage value, there is no way for clients to know the value behind the key (now they get it from metadata as constant).

I don't 100% get what you mean but I will just say my implementation don't have this issue.

**Comment by muharem** [comment]
true, solved with `EnsureOriginWithArg`

regarding the last point, let me give you an example. 
```
pallet_salary::Config{ 
   ...
   #[pallet::constant]
   Budget: Get<Balance>
}
```
now with this configuration and static value for `Budget` clients can get this value from the metadata.
what if I replace that static value by a getter like 
```
impl<Default, ParameterStore> Get<..> for Budget<..> {
  fn get() -> Balance {
      ParameterStore::get(key).unwrap_or_else(Default::get())
  }
}
```
how clients will retrieve the budget?

**Comment by bkchr** [comment]
> Basically replace `api.const.salary.budge` with `api.query.parameters.parameters({ Salary: { Budget: undefiled } })`

I like :)
